src/hooks/core/users/useFollowUserV2Hook.tsx

```diff
@@ -0,0 +1,135 @@
+
+    try {
+      const result = await checkFollowStatus(channelId);
+      StorageUtils.otherProfileData.setForKey(channelId, JSON.stringify(result));
+    } catch (e) {
+      console.error('get follow status failed', e);
+    }
+  }
+
+  const getOtherProfileFollowStatus = async (channelId: string | undefined): Promise<any> => {
+    if (!channelId) return;
+
+    let otherProfileDataCache = StorageUtils.otherProfileData.getForKey(channelId);
+    let otherProfileDataJSON: FollowStatusAPI = {
+      isMeFollowingTargetFromAction: undefined,
+      isMeFollowingTarget: undefined,
+      isTargetFollowingMe: undefined
+    };
+
+    try {
+      otherProfileDataJSON = JSON.parse(otherProfileDataCache || '{}');
+    } catch (e) {
+      console.error(e);
+    }
+
+    if (otherProfileDataCache) {
+      setFollowStatus({
+        isFollowed: otherProfileDataJSON?.isTargetFollowingMe,
+        isFollowing: otherProfileDataJSON?.isMeFollowingTarget,
+        isFollowingFromAction: otherProfileDataJSON?.isMeFollowingTargetFromAction
+      });
+    }
+
+    await __helperCallOtherProfileData(channelId);
+    otherProfileDataCache = StorageUtils.otherProfileData.getForKey(channelId);
+    try {
+      otherProfileDataJSON = JSON.parse(otherProfileDataCache || '{}');
+    } catch (e) {
+      console.error(e);
+    }
+
+    setFollowStatus({
+      isFollowed: otherProfileDataJSON?.isTargetFollowingMe,
+      isFollowing: otherProfileDataJSON?.isMeFollowingTarget,
+      isFollowingFromAction: otherProfileDataJSON?.isMeFollowingTargetFromAction
+    });
+  };
+
+  const followUserAction = async (item: ChannelList) => {
+    /**
+     * Optimistic state update (START)
+     */
+    const initialFollowStatus = {...followStatus};
+    setFollowStatus((prevFollowStatus) => ({
+      ...prevFollowStatus,
+      isFollowingFromAction: !initialFollowStatus?.isFollowingFromAction,
+      isFollowing: !initialFollowStatus?.isFollowing
+    }));
+
+    /**
+     * Optimistic state update (END)
+     */
+
+    try {
+      const myId = tab === 'signed' ? signedProfileId : anonProfileId;
+      const targetUser = item?.rawJson?.channel?.members?.find(
+        (member) => member?.user_id !== myId
+      )?.user;
+
+      const data = {
+        user_id_followed: targetUser?.id,
+        follow_source: 'chat'
+      };
+
+      if (initialFollowStatus?.isFollowing) {
+        await setUnFollow(data);
+      } else {
+        await setFollow(data);
+      }
+      __helperCallOtherProfileData(item?.id);
+    } catch (e) {
+      const action = initialFollowStatus?.isFollowing ? 'unfollow' : 'follow';
+      SimpleToast.show(`Failed to ${action} user. Please try again later`);
+      /**
+       * Revert optimistic update
+       */
+      setFollowStatus(() => ({
+        ...initialFollowStatus
+      }));
+    }
+  };
+
+  return {
+    followStatus,
+
+    followUserAction,
+    getOtherProfileFollowStatus
+  };
+};
+
+export default useFollowUserV2Hook;
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The `useFollowUserV2Hook` hook seems to be well implemented. It provides a clear separation of concerns, with each function having a specific purpose. The use of optimistic UI updates is also a good practice for improving user experience. However, there are some areas that could be improved:

1. Error handling: Currently, the error messages are being logged to the console and shown as a toast notification. While this might be sufficient in some cases, it would be better to have a more robust error handling mechanism. For instance, you could consider propagating the errors up to the components using this hook and let them decide how to handle these errors.

2. Caching: The hook is currently using local storage for caching the follow status. This can lead to stale data if the follow status changes elsewhere (e.g., in another tab or device). Consider using a state management library like Redux or MobX, which can help manage global state across the application and keep the data consistent.

3. Async operations: There are several async operations in this hook, but there's no way to cancel them if the component using this hook unmounts before they complete. This can lead to memory leaks and unexpected behavior. Consider using an effect cleanup function to cancel these operations when they're no longer needed.

4. Code duplication: There's some code duplication in the `getOtherProfileFollowStatus` function where you're setting the follow status twice. You could refactor this to avoid repeating the same logic.

Here's a suggestion on how you could address these points:

```diff
  // ...existing code...

  const getOtherProfileFollowStatus = async (channelId: string | undefined): Promise<any> => {
    if (!channelId) return;

    let otherProfileDataCache = StorageUtils.otherProfileData.getForKey(channelId);
    let otherProfileDataJSON: FollowStatusAPI = {
      isMeFollowingTargetFromAction: undefined,
      isMeFollowingTarget: undefined,
      isTargetFollowingMe: undefined
    };

    try {
      otherProfileDataJSON = JSON.parse(otherProfileDataCache || '{}');
    } catch (e) {
      console.error(e);
    }

-   if (otherProfileDataCache) {
-     setFollowStatus({
-       isFollowed: otherProfileDataJSON?.isTargetFollowingMe,
-       isFollowing: otherProfileDataJSON?.isMeFollowingTarget,
-       isFollowingFromAction: otherProfileDataJSON?.isMeFollowingTargetFromAction
-     });
-   }

    await __helperCallOtherProfileData(channelId);
    otherProfileDataCache = StorageUtils.otherProfileData.getForKey(channelId);
    try {
      otherProfileDataJSON = JSON.parse(otherProfileDataCache || '{}');
    } catch (e) {
      console.error(e);
    }

    setFollowStatus({
      isFollowed: otherProfileDataJSON?.isTargetFollowingMe,
      isFollowing: otherProfileDataJSON?.isMeFollowingTarget,
      isFollowingFromAction: otherProfileDataJSON?.isMeFollowingTargetFromAction
    });
  };

  // ...existing code...
```

This change removes the first `setFollowStatus` call, which was unnecessary because you're updating the follow status again after fetching the latest data.


<!-- This is an auto-generated comment by OSS CodeRabbit -->